rust/src/detect/xbits.rs

```diff
@@ -0,0 +1,224 @@
+                    cmd: get_xbit_cmd("noalert")?,
+                    name: std::ptr::null_mut(),
+                    tracker: 0,
+                    expire: 0,
+                    vartype: 0,
+                })
+            }
+            _ => {
+                SCLogError!("xbits noalert parsing failed");
+                return Err(());
+            }
+        },
+        3 | 4 => {
+            res = match evaluate_args(split_args) {
+                Ok(val) => val,
+                Err(_) => return Err(()),
+            };
+        }
+        e => {
+            SCLogError!("Erroneous number of arguments: {}", e);
+            return Err(());
+        }
+    }
+    Ok(RSXBitsData {
+        cmd: res.0,
+        name: (res.1).into_raw(),
+        tracker: res.2,
+        expire: res.3,
+        vartype: res.4,
+    })
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_xbits_parse(carg: *const c_char) -> *mut c_void {
+    if carg.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let arg = match CStr::from_ptr(carg).to_str() {
+        Ok(arg) => arg,
+        _ => {
+            return std::ptr::null_mut();
+        }
+    };
+
+    match parse_xbits(arg) {
+        Ok(detect) => Box::into_raw(Box::new(detect)) as *mut _,
+        Err(_) => {
+            return std::ptr::null_mut();
+        }
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn rs_xbits_free(ptr: *mut c_void, name: *const c_char) {
+    if !ptr.is_null() {
+        let _str = CString::from_raw(name as *mut c_char);
+        std::mem::drop(Box::from_raw(ptr as *mut RSXBitsData));
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_bad_input() {
+        assert_eq!(true, parse_xbits("alert").is_err());
+        assert_eq!(true, parse_xbits("n0alert").is_err());
+        assert_eq!(true, parse_xbits("nOalert").is_err());
+        assert_eq!(
+            true,
+            parse_xbits("set,abc,track nonsense, expire 3600").is_err()
+        );
+        assert_eq!(
+            true,
+            parse_xbits("set,abc,track ip_source, expire 3600").is_err()
+        );
+        assert_eq!(
+            true,
+            parse_xbits("set,abc,track ip_src, expire -1").is_err()
+        );
+        assert_eq!(true, parse_xbits("set,abc,track ip_src, expire 0").is_err());
+        assert_eq!(
+            true,
+            parse_xbits("set,abc,expire 1000, track ip_dst").is_err()
+        );
+        assert_eq!(
+            true,
+            parse_xbits("set,abc,tracker ip_src, expire 0").is_err()
+        );
+    }
+
+    #[test]
+    fn test_good_input() {
+        assert_eq!(true, parse_xbits("noalert").is_ok());
+        assert_eq!(true, parse_xbits("set,abc,track ip_pair").is_ok());
+        assert_eq!(
+            true,
+            parse_xbits("set, abc ,track ip_pair, expire 3600").is_ok()
+        );
+        assert_eq!(
+            true,
+            parse_xbits("set  ,abc,track ip_src, expire 1234").is_ok()
```

Comment:
Should test that the parsed values are as expected.. No leading/trailing white space, etc.